ext/gd/libgd/gd_avif.c

```diff
@@ -0,0 +1,602 @@
+	(alpha7Bit == 127 ? \
+				0 : \
+				255 - ((alpha7Bit << 1) + (alpha7Bit >> 6)))
+
+#define alpha8BitTo7Bit(alpha8Bit) (gdAlphaMax - (alpha8Bit >> 1))
+
+
+/*** Helper functions ***/
+
+/* Convert the quality param we expose to the quantity params used by libavif.
+	 The *Quantizer* params values can range from 0 to 63, with 0 = highest quality and 63 = worst.
+	 We make the scale 0-100, and we reverse this, so that 0 = worst quality and 100 = highest.
+
+	 Values below 0 are set to 0, and values below MAX_QUALITY are set to MAX_QUALITY.
+*/
+static int quality2Quantizer(int quality) {
+	int clampedQuality = CLAMP(quality, 0, MAX_QUALITY);
+
+	float scaleFactor = (float) AVIF_QUANTIZER_WORST_QUALITY / (float) MAX_QUALITY;
+
+	return round(scaleFactor * (MAX_QUALITY - clampedQuality));
+}
+
+/*
+	 As of February 2021, this algorithm reflects the latest research on how many tiles
+	 and threads to include for a given image size.
+	 This is subject to change as research continues.
+
+	 Returns false if there was an error, true if all was well.
+ */
+static avifBool setEncoderTilesAndThreads(avifEncoder *encoder, avifRGBImage *rgb) {
+	int imageArea, tiles, tilesLog2, encoderTiles;
+
+	// _gdImageAvifCtx(), the calling function, checks this operation for overflow
+	imageArea = rgb->width * rgb->height;
+
+	tiles = (int) ceil((double) imageArea / MIN_TILE_AREA);
+	tiles = MIN(tiles, MAX_TILES);
+	tiles = MIN(tiles, MAX_THREADS);
+
+	// The number of tiles in any dimension will always be a power of 2. We can only specify log(2)tiles.
+
+	tilesLog2 = floor(log2(tiles));
+
+	// If the image's width is greater than the height, use more tile columns
+	// than tile rows to make the tile size close to a square.
+
+	if (rgb->width >= rgb->height) {
+		encoder->tileRowsLog2 = tilesLog2 / 2;
+		encoder->tileColsLog2 = tilesLog2 - encoder->tileRowsLog2;
+	} else {
+		encoder->tileColsLog2 = tilesLog2 / 2;
+		encoder->tileRowsLog2 = tilesLog2 - encoder->tileColsLog2;
+	}
+
+	// It's good to have one thread per tile.
+	encoderTiles = (1 << encoder->tileRowsLog2) * (1 << encoder->tileColsLog2);
+	encoder->maxThreads = encoderTiles;
+
+	return AVIF_TRUE;
+}
+
+/*
+	 We can handle AVIF images whose color profile is sRGB, or whose color profile isn't set.
+*/
+static avifBool isAvifSrgbImage(avifImage *avifIm) {
+	return
+		(avifIm->colorPrimaries == AVIF_COLOR_PRIMARIES_BT709 ||
+			avifIm->colorPrimaries == AVIF_COLOR_PRIMARIES_UNSPECIFIED) &&
+		(avifIm->transferCharacteristics == AVIF_TRANSFER_CHARACTERISTICS_SRGB ||
+			avifIm->transferCharacteristics == AVIF_TRANSFER_CHARACTERISTICS_UNSPECIFIED)
+	;
+}
+
+/*
+	 Check the result from an Avif function to see if it's an error.
+	 If so, decode the error and output it, and return true.
+	 Otherwise, return false.
+*/
+static avifBool isAvifError(avifResult result, const char *msg) {
+	if (result != AVIF_RESULT_OK) {
+		zend_error(E_ERROR, "avif error - %s: %s", msg, avifResultToString(result));
```

Comment:
This should probably be using E_WARNING or E_NOTICE, not a fatal error.

Reply:
Why no `gd_error()` as [in upstream](https://github.com/libgd/libgd/blob/82d260950589563a1af9c56f4ce5fde843a695ae/src/gd_avif.c#L152-L159)?

Reply:
Which do you think is best?

I was inspired by the `E_ERROR` errors in comparable situations in `gd_webp.c`. But, for all I know, perhaps those should be non-fatal as well :)

Reply:
IMO, the fewer differences from upstream libgd, the better.

Reply:
ah, @cmb69 , now I see your comment too.

Happy to use `gd_error()` instead, which I do also see in the PHP repo.